server/src/main/java/io/crate/replication/logical/action/PublicationsStateAction.java

```diff
@@ -0,0 +1,274 @@
+
+    public PublicationsStateAction() {
+        super(NAME);
+    }
+
+    @Override
+    public Writeable.Reader<Response> getResponseReader() {
+        return Response::new;
+    }
+
+    @Singleton
+    public static class TransportAction extends TransportMasterNodeReadAction<Request, Response> {
+
+        private final LogicalReplicationService logicalReplicationService;
+        private final Schemas schemas;
+
+        @Inject
+        public TransportAction(TransportService transportService,
+                               ClusterService clusterService,
+                               ThreadPool threadPool,
+                               IndexNameExpressionResolver indexNameExpressionResolver,
+                               LogicalReplicationService logicalReplicationService,
+                               Schemas schemas) {
+            super(Settings.EMPTY,
+                  NAME,
+                  false,
+                  transportService,
+                  clusterService,
+                  threadPool,
+                  indexNameExpressionResolver,
+                  Request::new);
+            this.logicalReplicationService = logicalReplicationService;
+            this.schemas = schemas;
+
+            TransportActionProxy.registerProxyAction(transportService, NAME, Response::new);
+        }
+
+        @Override
+        protected String executor() {
+            return ThreadPool.Names.SAME;
+        }
+
+        @Override
+        protected Response read(StreamInput in) throws IOException {
+            return new Response(in);
+        }
+
+        @Override
+        protected void masterOperation(Request request,
+                                       ClusterState state,
+                                       ActionListener<Response> listener) throws Exception {
+            List<RelationDetails> relationDetails = new ArrayList<>();
+            for (var publicationName : request.publications()) {
+                var publication = logicalReplicationService.publications().get(publicationName);
+                if (publication == null) {
+                    listener.onFailure(new PublicationUnknownException(publicationName));
+                    return;
+                }
+
+                List<RelationName> relationNames = resolveRelationsNames(publication, schemas);
+                for (var relationName : relationNames) {
+                    var relationDetail = relationDetail(relationName, state);
+                    if (relationDetail != null) {
+                        relationDetails.add(relationDetail);
+                    }
+                }
+            }
+            var response = new Response(
+                relationDetails.stream().map(d -> d.relationName).toList(),
+                relationDetails.stream().flatMap(d -> d.concreteIndices.stream()).toList(),
+                relationDetails.stream().flatMap(d -> d.concreteTemplates.stream()).toList()
```

Comment:
Maybe premature optimization, but instead of building up `relationDetails` just to convert it into 3 other lists, the loop above could build the three lists directly.


Or maybe create a Response builder? Looks like the entire purpose of the `RelationDetails` is to help constructing this response.